bcgov/design-system
# Jest test framework for React components

This PR adds tests to the React component library using Jest with [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/). A minimal series of tests for the Button component is included to show the test framework in action.

## Why Jest

I've chosen Jest as the testing framework because of its maturity and integration with the React testing ecosystem. I tested using [Node's new native test runner](https://nodejs.org/docs/latest-v20.x/api/test.html), but found I couldn't get it to read TypeScript files properly with `ts-node` or similar solutions. In testing with plain JavaScript files, `node:test` seems to run much faster than Jest. Node's test runner lacks the [syntactic sugar of Jest's `expect` API](https://jestjs.io/docs/expect), so you end up writing fairly verbose native `assert` code unless you opt to [install `expect` without Jest](https://www.npmjs.com/package/expect). In the future, we might be able to remove the Jest dependency from our testing suite if a good native Node pattern emerges for React component testing.

## Why React Testing Library

I opted for React Testing Library for its DOM testing features, like firing events based on a synthetic click event. It seems to be a de facto standard for React unit testing [since Enyzme stopped getting regular updates](https://piotrstaniow.pl/goodbye-enzyme). This could work alongside something like Cypress in the future if we have more involved end-to-end tests, or it could be replaced by Cypress's own DOM testing features.

## Scripts

- `npm run test` does a standard run of the test suite
- `npm run test:ci` passes [Jest's `--ci` flag](https://jestjs.io/docs/cli#--ci) that changes how snapshots are treated in a continuous integration environment (like GitHub Actions)
- `npm run test:watch` uses [Jest's `--watch` flag](https://jestjs.io/docs/cli#--watch) to keep the tests running during local development

## How to test

1. Pull this branch and run `npm install` in the `packages/react-components` directory to install all of the new test dependencies.
2. Run `npm run test` to run the test suite. If everything is working, the output should be similar to below:

<img width="827" alt="Jest test run output" src="https://github.com/bcgov/design-system/assets/25143706/48e86bf2-e428-4f67-ae49-95b58620d6b8">

```
➜ npm run test

> @bcgov/design-system-react-components@0.0.6 test
> npx jest --config=./jest.config.ts

ts-jest[config] (WARN) message TS151001: If you have issues related to imports, you should consider setting `esModuleInterop` to `true` in your TypeScript configuration file (usually `tsconfig.json`). See https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/#easier-ecmascript-module-interoperability for more information.
 PASS  src/components/Button/Button.test.tsx (19.255 s)
  renders the component correctly
    ✓ appears in the document (3 ms)
    ✓ is enabled by default (1 ms)
  handles a click event with onPress prop
    ✓ has been clicked
  can be disabled with isDisabled prop
    ✓ is disabled with prop passed (1 ms)

Test Suites: 1 passed, 1 total
Tests:       4 passed, 4 total
Snapshots:   0 total
Time:        24.259 s
Ran all test suites.
```

Note that I'm leaving the `tsconfig.json` file's `esModuleInterop` property as-is for now despite the warning in the Jest output since the tests are running as expected.
